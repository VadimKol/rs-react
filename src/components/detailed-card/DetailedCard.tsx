import { type ReactNode, useEffect, useState } from 'react';
import { useNavigate, useOutletContext } from 'react-router-dom';
import { type Character, getCharacter } from 'rickmortyapi';

import { CustomButton } from '../custom-button/Ð¡ustomButton';
import { ImageBlock } from '../image-block/ImageBlock';
import styles from './styles.module.scss';

export function DetailedCard(): ReactNode {
  const { characterID, setNoMatch } = useOutletContext<{
    characterID: string;
    setNoMatch: (noMatch: boolean) => void;
  }>();
  const [character, setCharacter] = useState<Character>({} as Character);
  const navigate = useNavigate();
  const [err, setErr] = useState('');
  const [loader, setLoader] = useState(true);

  if (err) {
    throw new Error(err);
  }

  useEffect(() => {
    setLoader(true);
    getCharacter(Number(characterID))
      .then((response) => {
        if (response.status === 200) {
          setCharacter(response.data);
        } else if (response.status === 404) {
          setNoMatch(true);
        } else {
          throw new Error(response.statusMessage);
        }
      })
      .catch((error: Error) => setErr(error.message))
      .finally(() => setLoader(false));
  }, [characterID, setNoMatch]);

  const desc = [
    `Species: ${character?.species}`,
    `Gender: ${character?.gender}`,
    `Status: ${character?.status}`,
    `Origin: ${character?.origin?.name}`,
    `Location: ${character?.location?.name}`,
  ];

  return (
    <section className={styles.detailed_Card}>
      {loader ? (
        <div className={styles.loader} />
      ) : (
        <>
          <div className={styles.card}>
            <ImageBlock src={character?.image || ''} alt="Character" />
            <div className={styles.text_container}>
              <h2 className={styles.card_title}>{character?.name}</h2>
              <ul className={styles.desc}>
                {desc.map((item) => (
                  <li key={item} className={styles.desc_item}>
                    {item}
                  </li>
                ))}
              </ul>
              {character?.type !== '' && <p className={styles.type}>Type: {character?.type}</p>}
              <p className={styles.episodes}>
                Episodes: {character?.episode.map((episode) => episode.replace(/\D/g, '')).join(', ')}
              </p>
            </div>
          </div>
          <CustomButton className={styles.close} onClick={() => navigate('/')}>
            Close
          </CustomButton>
        </>
      )}
    </section>
  );
}
