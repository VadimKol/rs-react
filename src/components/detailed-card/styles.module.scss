.detailed_Card {
  display: flex;
  flex-direction: column;
  gap: 20px;
  align-items: center;
}

.card {
  position: relative;
  width: 265px;
  background:
    linear-gradient(var(--background-color) 0 0) padding-box,
    linear-gradient(to top, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;
}

.text_container {
  display: flex;
  flex-direction: column;
  gap: 10px;
  padding: 20px 30px;
  box-shadow: inset 0 5px 5px 0 #d3d3d3;
}

.card_title {
  margin: 0;
  font-size: 18px;
  font-weight: 800;
  text-align: center;
}

.desc {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.desc_item,
.type {
  margin: auto;
}

.episodes {
  display: -webkit-box;
  overflow: hidden;
  margin: 0 auto;
  text-align: left;
  text-overflow: ellipsis;
  -webkit-line-clamp: 5;
  -webkit-box-orient: vertical;
}

.close {
  font-size: 1.3rem;
}

.loader {
  width: 265px;
  height: 265px;
  background-image: url("/src/assets/images/loader.png");
  background-position: center;
  background-size: contain;
  background-repeat: no-repeat;
  animation: loader 1s infinite linear;
}

@keyframes loader {
  to {
    transform: rotate(1turn);
  }
}
