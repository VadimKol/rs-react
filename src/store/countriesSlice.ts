import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { countries } from '@/common/countries';

import { type RootState } from './store';

export const countriesLice = createSlice({
  name: 'countries',
  initialState: countries,
  reducers: {
    addCountry: (state, action: PayloadAction<string>) => {
      state.push(action.payload);
    },
  },
});

export const { addCountry } = countriesLice.actions;

export const selectCountries = (state: RootState): string[] => state.countries;

export const useCountries = (): string[] => useSelector(selectCountries);
