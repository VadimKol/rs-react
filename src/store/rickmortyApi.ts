import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react';
import { getCharacter, getCharacters } from 'rickmortyapi';

export const rickmortyApi = createApi({
  reducerPath: 'rickmortyApi',
  baseQuery: fakeBaseQuery(),
  endpoints: (builder) => ({
    getCharacters: builder.query({
      queryFn: async ({ page, character }: { page: number; character: { name: string } }) => {
        try {
          const { data, status, statusMessage } = await getCharacters({ page, name: character.name });

          if (status === 200) {
            return { data: { characters: data.results || [], totalPages: data.info?.pages || 0 } };
          }
          if (status === 404) {
            return { data: { characters: [], totalPages: 0 } };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
    getCharacter: builder.query({
      queryFn: async (id: number) => {
        try {
          const { data, status, statusMessage } = await getCharacter(id);

          if (status === 200) {
            return { data };
          }
          if (status === 404) {
            return { data: null };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
  }),
});

export const { useGetCharactersQuery, useGetCharacterQuery } = rickmortyApi;
