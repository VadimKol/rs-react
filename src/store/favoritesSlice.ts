/* import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import { type Character } from 'rickmortyapi';

import { type RootState } from './store';

const initialState: Character[] = [];

export const favoritesSLice = createSlice({
  name: 'favorites',
  initialState,
  reducers: {
    addToFavorites: (state, action: PayloadAction<Character>) => {
      state.push(action.payload);
    },
    removeFromFavorites: (state, action: PayloadAction<Character>) =>
      state.filter((character) => character.id !== action.payload.id),
    clearFavorites: () => [],
  },
});

export const { addToFavorites, removeFromFavorites, clearFavorites } = favoritesSLice.actions;

const selectFavorites = (state: RootState): Character[] => state.favorites;

const isFavorite = (state: RootState, character?: Character): boolean =>
  state.favorites.some((favorite) => favorite.id === character?.id);

export const useFavorites = (): Character[] => useSelector(selectFavorites);

export const useIsFavorite = (character?: Character): boolean =>
  useSelector((state: RootState) => isFavorite(state, character));
 */
