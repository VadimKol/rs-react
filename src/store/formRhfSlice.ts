import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

const initialState: FormInfo = {
  name: null,
  age: null,
  email: null,
  password: null,
  confirmPassword: null,
  gender: null,
  country: null,
  tc: null,
  image: null,
  strength: '0',
};

export const formRhfSLice = createSlice({
  name: 'formRhf',
  initialState,
  reducers: {
    updateFormRhf: (_, action: PayloadAction<FormInfo>) => action.payload,
  },
});

export const { updateFormRhf } = formRhfSLice.actions;

const selectFormRhfInfo = (state: RootState): FormInfo => state.formRhf;

export const useFormRhfInfo = (): FormInfo => useSelector(selectFormRhfInfo);
