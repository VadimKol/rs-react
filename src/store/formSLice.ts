import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

export const initialState: FormInfo[] = [];

export const formSLice = createSlice({
  name: 'form',
  initialState,
  reducers: {
    updateForm: (state, action: PayloadAction<FormInfo>) => {
      state.push(action.payload);
    },
  },
});

export const { updateForm } = formSLice.actions;

const selectFormInfo = (state: RootState): FormInfo[] => state.form;

export const useFormInfo = (): FormInfo[] => useSelector(selectFormInfo);
