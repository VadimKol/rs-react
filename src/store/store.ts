import { configureStore } from '@reduxjs/toolkit';

import { formSLice } from './formSLice';

export const store = configureStore({ reducer: { form: formSLice.reducer } });

export type RootState = ReturnType<typeof store.getState>;
