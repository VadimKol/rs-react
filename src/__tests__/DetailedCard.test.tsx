import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { DetailedCard } from '@/components/detailed-card/DetailedCard';
import { StoreProvider } from '@/store/StoreProvider';

import { character } from './__mocks__/data';
import { mockRouterPush } from './setupAfterEnv';

describe('DetailedCard Component', () => {
  const user = userEvent.setup();

  it('renders correctly', () => {
    const { container } = render(
      <StoreProvider>
        <DetailedCard character={character} />
      </StoreProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('changes URL when click on Close button', async () => {
    render(
      <StoreProvider>
        <DetailedCard character={character} />
      </StoreProvider>,
    );

    const buttonSearch = screen.getByText('Close');

    await user.click(buttonSearch);

    expect(mockRouterPush).toHaveBeenCalledWith(`/?page=1&search=`);
  });

  it('did not render when person is not specified', () => {
    const { container } = render(
      <StoreProvider>
        <DetailedCard character={null} />
      </StoreProvider>,
    );

    expect(container).toBeEmptyDOMElement();
  });
});
