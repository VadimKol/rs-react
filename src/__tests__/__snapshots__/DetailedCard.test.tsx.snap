// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailedCard Component renders correctly 1`] = `
<div>
  <section
    class="detailed_Card"
  >
    <div
      class="loader"
    />
  </section>
</div>
`;
